### PR TITLE
Release R136

### PR DESCRIPTION
**Pull request by dilyand**

### CODE REVIEW

**Comment by snowplowcla** [comment]
Thanks for your pull request. Is this your first contribution to a Snowplow open source project? Before we can look at your pull request, you'll need to sign a Contributor License Agreement (CLA).

:memo: Please visit https://docs.snowplowanalytics.com/docs/contributing/contributor-license-agreement/ to learn more and sign.

Once you've signed, please reply here (e.g. I signed it!) and we'll verify. Thanks.

File: schemas/nl.basjes/yauaa_context/jsonschema/1-0-4
Comment (istreeter) [inline schemas/nl.basjes/yauaa_context/jsonschema/1-0-4:18]: Why the change from type string to type `[string, null]`?  I'm not against it if you have a good reason.  It struck me as odd because enrich doesn't output nulls.

File: schemas/nl.basjes/yauaa_context/jsonschema/1-0-4
Comment (istreeter) [inline schemas/nl.basjes/yauaa_context/jsonschema/1-0-4:44]: You added "null" as a type... but this does not make any difference, unless you also add `null` to the enum array.  As currently written, null is still not an allowed value for `operatingSystemClass`.

**Review by istreeter — COMMENTED** [review]
Hey @dilyand please can you explain more about what you are trying to achieve with this schema change, and therefore why you chose these values for `maxLength`?

We spoke earlier about this being "the schema change to end all schema changes" -- meaning we're all fed up of updating enrich each time the yauaa enrichment sees an unexpected user agent string.  So we talked about lengthening each field to protect it from future changes.

But with that in mind, I can see you have chosen a mixture of 256, 1000, and 1024 as max length for different fields.  How did you decide on these values?

File: schemas/nl.basjes/yauaa_context/jsonschema/1-0-4
Comment (dilyand) [reply schemas/nl.basjes/yauaa_context/jsonschema/1-0-4:18]: They are optional fields, so presumably at some point they may be sent with a `null` value. We even have it as a check in igluctl. 

I appreciate we currently do not do that in enrich, but if we change it in the future, we won't need to worry about the schema not allowing it. It's part of trying to make the schema more future-proof.

File: schemas/nl.basjes/yauaa_context/jsonschema/1-0-4
Comment (dilyand) [reply schemas/nl.basjes/yauaa_context/jsonschema/1-0-4:44]: Yes, I did not realise that. I think if we keep null as an allowed type, I'll add it to the enum.

**Comment by dilyand** [comment]
> Hey @dilyand please can you explain more about what you are trying to achieve with this schema change, and therefore why you chose these values for `maxLength`?
> 
> We spoke earlier about this being "the schema change to end all schema changes" -- meaning we're all fed up of updating enrich each time the yauaa enrichment sees an unexpected user agent string. So we talked about lengthening each field to protect it from future changes.
> 
> But with that in mind, I can see you have chosen a mixture of 256, 1000, and 1024 as max length for different fields. How did you decide on these values?

I put most of the reasoning in the issue #1216, but very briefly:
- All fields that are extracted from the useragent string have a limit of 1000 because this is the max length for the useragent field in atomic.events. So, even if the whole string is interpreted as a single property, it still can't exceed that value.
- All fields that are deduced (rather than extracted) have a limit of 128 / 256. This somewhat arbitrary. I just wanted to increase the existing limits which seemed a bit too strict. My thinking here is that these values are controlled by the YAUAA library, not by whoever puts together the useragent strings, and so are less likely to fluctuate widely.
- The fields with length 1024 were already like that. I do not want to lower this limit because that is not a backward-compatible change. But in practice, we should never have anything higher than 1000 because of the limitation for the atomic useragent field.

File: schemas/nl.basjes/yauaa_context/jsonschema/1-0-4
Comment (istreeter) [reply schemas/nl.basjes/yauaa_context/jsonschema/1-0-4:18]: My gut feeling at the moment is like this:

If you're going to allow nulls for some fields, then you need to allow nulls for all fields, or else it's not consistent between fields.  Which means allow nulls for the enum fields too, like `operatingSystemClass`.  Which means add `null` to the enum array.

If you're adding `null` to the enum array then you need to be very very sure that RDB Loader will not choke on the schema migration.  Remember that RDB Loader *immediately* (and probably wrongly) tries to migrate the table as soon as the schema is published; it does not wait until it sees the first `1-0-4` event.  Remember also, we need to be sure that old versions of RDB Loader also don't choke on the schema migration, otherwise we might break oss pipelines just by publishing a schema.

Now RDB Loader _probably_ handles this OK.  I'm just saying we need to be 100% sure before merging this.

For that reason.... personally I would avoid adding the nulls, and put it back to be more like how it was in 1-0-3.  There is no reason whatsoever to add nulls to this schema.  I see no reason why enrich will ever change in future and to start emitting nulls.

Yes we want to make the schema future-proof, but I see it as future-proofing against things we are not in control of, like crazy user agent strings.  Whereas nulls are something we are completely in control of, so we don't need to future-proof against nulls.

Comment (dilyand) [reply schemas/nl.basjes/yauaa_context/jsonschema/1-0-4:18]: I think that is a sensible idea. I was trying to make igluctl happy, but I agree that in this instance, where we're in control of this, there is no need to worry.

File: schemas/nl.basjes/yauaa_context/jsonschema/1-0-4
Comment (istreeter) [inline schemas/nl.basjes/yauaa_context/jsonschema/1-0-4:3]: Entity context or context entity?

**Review by istreeter — APPROVED** [review]
Looks good.

**Comment by oguzhanunlu** [comment]
yauaa 7.4.0 has `WebviewAppNameVersion` as well, it is not documented [here](https://yauaa.basjes.nl/using/memoryusage/)

[ref](https://github.com/nielsbasjes/yauaa/blob/v7.4.0/analyzer/src/main/java/nl/basjes/parse/useragent/UserAgent.java#L120)

is there a specific reason we didn't include `webviewAppNameVersion` field in this new version of the schema?

**Comment by oguzhanunlu** [comment]
I added a separate commit to add `webviewAppNameVersion` cc @istreeter @dilyand

**Comment by adatzer** [comment]
Could we also merge https://github.com/snowplow/iglu-central/pull/1221 into `r136`? cc @oguzhanunlu @dilyand

File: CHANGELOG
Comment (paulboocock) [inline CHANGELOG:6]: nit: I don't think we usually have the word `close` in these lines

File: schemas/nl.basjes/yauaa_context/jsonschema/1-0-4
Comment (paulboocock) [reply schemas/nl.basjes/yauaa_context/jsonschema/1-0-4:3]: "context entity" feels more right to me. I'd probably just go for "entity" though.

**Review by paulboocock — APPROVED** [review]
LGTM. Couple of nit pick comments but happy for this to go out once the changelog is tidied.

File: schemas/nl.basjes/yauaa_context/jsonschema/1-0-4
Comment (dilyand) [reply schemas/nl.basjes/yauaa_context/jsonschema/1-0-4]: I changed it back to just 'context'. I think the word entity confuses things in this case. This schema does not describe an entity, but rather just a grab bag of things that could be deduced from the useragent.

File: CHANGELOG
Comment (dilyand) [reply CHANGELOG]: Done.

**Review by paulboocock — APPROVED** [review]

File: CONTRIBUTING.md
Comment (jbeemster) [inline CONTRIBUTING.md:22]:
```suggestion
[cla]: https://docs.snowplow.io/docs/contributing/contributor-license-agreement/
```

File: schemas/com.snowplowanalytics.mobile/remote_config/jsonschema/1-0-1
Comment (jbeemster) [inline schemas/com.snowplowanalytics.mobile/remote_config/jsonschema/1-0-1:1]: @istreeter doing a diff with 1-0-0 a lot of these fields now have a `maximum` as well as a few now allowing `object, null` pairing over just `object`.  Just want to confirm that the addition of new limits and type groupings is not going to cause issues in any of our loaders / data-warehouses?

I am not entirely certain if this schema gets tracked into pipelines to be fair but if it was would these changes get automatically represented in their respective columns?

Comment (matus-tomlein) [reply schemas/com.snowplowanalytics.mobile/remote_config/jsonschema/1-0-1:1]: The question is still valid, just wanted to say that the schema is not being used in pipelines, it is just to help developers better define their configurations (some IDEs may use it for validating the JSONs).

File: README.md
Comment (adatzer) [reply README.md]: All suggested changes applied to contributing and readme files:
 - Links to snowplow.io
 - "Snowplow Analytics" to "Snowplow"

File: schemas/com.snowplowanalytics.mobile/remote_config/jsonschema/1-0-1
Comment (istreeter) [reply schemas/com.snowplowanalytics.mobile/remote_config/jsonschema/1-0-1:1]: I tested this on a pipeline with a loader.  Even if it's not expected to be used in tracking, it was worth testing.

It is completely safe, because the field `configurationBundle` is a complex object, so RDB transformer simply stringifies the whole field, and RDB Loader loads it as a string.  Any change you make to a sub-field (e.g. `configurationBundle.namespace`) does not bother the transformer/loader at all.

File: schemas/nl.basjes/yauaa_context/jsonschema/1-0-4
Comment (istreeter) [reply schemas/nl.basjes/yauaa_context/jsonschema/1-0-4:1]: **WARNING** -- this migration did not work when I tested in with Redshift RDB Loader.  The new field got added to the table, but the existing columns were not updated for the new lengths.

I opened [this issue in rdb loader](https://github.com/snowplow/snowplow-rdb-loader/issues/1051) to investigate it.  I spent quite a while trying to pinpoint the error, but no luck yet.

I'm not comfortable with merging this release until we have found why RDB Loader did not alter the lengths as it should.

File: schemas/com.snowplowanalytics.mobile/remote_config/jsonschema/1-0-1
Comment (istreeter) [inline schemas/com.snowplowanalytics.mobile/remote_config/jsonschema/1-0-1:41]: Hey @matus-tomlein just a warning, the "null" on this line does not do what you think it does!  If you want this field to be nullable then you must also add `null` to the enum:

```
"enum": ["get", "post", null]
```

Otherwise, a null will get rejected during validation.

I know this is slightly surprising.  You can test it out though [using an online validator](https://www.jsonschemavalidator.net/).

Comment (matus-tomlein) [reply schemas/com.snowplowanalytics.mobile/remote_config/jsonschema/1-0-1:41]: Thanks @istreeter, didn't know about this! Fixed.

File: schemas/io.snowplow.foundation/content/jsonschema/1-0-0
Comment (paulboocock) [inline schemas/io.snowplow.foundation/content/jsonschema/1-0-0:1]: If I'm being nitpicky, all these `io.snowplow.foundation` schemas are upsidedown compared to the rest of the Iglu Central schemas. I appreciate thats how the console renders them but would it be good to maintain Iglu Central patterns here?

File: schemas/io.snowplow.foundation/content/jsonschema/1-0-0
Comment (paulboocock) [inline schemas/io.snowplow.foundation/content/jsonschema/1-0-0:31]: What does `foundation` represent/mean? Are these not the recipe schemas from Try Snowplow? Why not `io.snowplow.recipe` or a final part that represents the "usecase" we typically see this schema used in `io.snowplow.content`?

File: schemas/io.snowplow.foundation/object/jsonschema/1-0-0
Comment (paulboocock) [inline schemas/io.snowplow.foundation/object/jsonschema/1-0-0:34]: This is a weird old schema. Not something we'd ever suggest someone uses in the real world as its so generic. Whats the use case for a schema like this?

**Review by paulboocock — COMMENTED** [review]
Technically I have no issues with the schemas, but there's a few nitpicks and a few questions on if these are the schemas we want to be commiting to for whatever use case they are being used to build (especially `object` 🤮)

File: schemas/io.snowplow.foundation/content/jsonschema/1-0-0
Comment (istreeter) [reply schemas/io.snowplow.foundation/content/jsonschema/1-0-0:31]: I was OK with the name "foundation" for the vendor, meaning like a collection of schemas that can be used when first starting off with Snowplow, before graduating to more advanced use cases when the user would want to design their own bespoke schemas.

But... I find the names of the schemas a bit strange.  The names "content" and "conversion" don't mean anything to me, but maybe that's because I don't come from the world of web analytics.

I don't have any problem with the schemas though, so I'd be OK to approve this release.

Comment (zephemeral) [reply schemas/io.snowplow.foundation/content/jsonschema/1-0-0:31]: `foundation` was the group consensus for what Ian describes

we didn't want to go with use-case specific names in case we later want to add more advanced content / conversion / funnel schemas that would then clash with these

Comment (zephemeral) [reply schemas/io.snowplow.foundation/content/jsonschema/1-0-0:31]: I'd be happy with `recipe` too, I don't who that decision is ultimately up to

File: schemas/io.snowplow.foundation/content/jsonschema/1-0-0
Comment (zephemeral) [reply schemas/io.snowplow.foundation/content/jsonschema/1-0-0:1]: potentially a dumb question but I'm comparing a couple of schemas and I don't know what upsidedown means – could you clarify? 
If there's a pattern we should follow then sure we can rewrite the schemas for consistency's sake

File: schemas/io.snowplow.foundation/object/jsonschema/1-0-0
Comment (zephemeral) [reply schemas/io.snowplow.foundation/object/jsonschema/1-0-0:34]: Good point, I just checked and it's not actually used in any of the tutorials, so I'm happy to ditch this one

File: schemas/io.snowplow.foundation/content/jsonschema/1-0-0
Comment (paulboocock) [reply schemas/io.snowplow.foundation/content/jsonschema/1-0-0]: Sorry to be that guy...

Now that I understand what `foundation` means, as in foundational... I think I prefer it over receipes as its more generic and allows these schemas to be used in a more broader context than just the recipe context.

**Review by jbeemster — APPROVED** [review]
LGTM - can I get a green-light when you lot get online and ill go ahead and merge it!

**Review by jbeemster — APPROVED** [review]

**Pull request merged**
